### PR TITLE
Fix for issue 32731

### PR DESCRIPTION
**Pull request by fadimounir**
Fix for #32731

Couple of fixes here:
1) Remove an assert in the TypeSystem's virtual function resolution (assert is too restrictive). Example: resolving `Base.Func` on an instance of `Derived`:
``` c#
class Base
{
    virtual void Func() { }
    virtual void MethodImplForFunc()
    {
        .override Base.Func()
    }
}
class Derived : Base
{
    override void MethodImplForFunc() { }
}
```
The assert was verifying that Base.Func was in the unification group of `Derived`. It would only be in the unification group of `Base` (where it gets removed later because of the methodImpl, and cause the assert at the derived type's level in the recursion)

2) Port some inlining rules from crossgen1 - One of them fixes an issue where we would incorrectly inline a virtual method that has a MethodImpl associated with it (test = self_override5)

3) Remove an assert from R2RDump related to the composite work (assert needs to be after loading _readyToRunHeaderRVA, and is actually already at the right place a few lines below)

### CODE REVIEW

**Comment by fadimounir** [comment]
@davidwrighton Please review the TypeSystem + inlining changes
@trylek Please review that small fix in the R2RDump tool (assert repeated twice, one of them at the wrong place)

cc @dotnet/crossgen-contrib

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs [hunk 1]
```diff
@@ -70,13 +70,36 @@ public abstract class Compilation : ICompilation
 
         public bool CanInline(MethodDesc caller, MethodDesc callee)
         {
+            if (JitConfigProvider.Instance.HasFlag(CorJitFlag.CORJIT_FLAG_DEBUG_CODE))
```
Comment (fadimounir) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: The changes here are ported from `CEEInfo::canInline` in jitinterface.cpp

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs [hunk 2]
```diff
@@ -70,13 +70,36 @@ public abstract class Compilation : ICompilation
...
+            }
+
             // Check to see if the method requires a security object.  This means they call demand and
             // shouldn't be inlined.
             if (callee.RequireSecObject)
             {
                 return false;
             }
 
+            // If the method is MethodImpl'd by another method within the same type, then we have
```
Comment (trylek) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: Nit - I have a vague recollection this corresponds to a Crossgen1 bit; could we provide a link to the relevant code for reference?

File: src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs [hunk 3]
```diff
@@ -381,7 +381,6 @@ private unsafe void Initialize(MetadataReader metadata)
                     throw new BadImageFormatException("The file is not a ReadyToRun image");
                 }
 
-                Debug.Assert(!Composite);
```
Comment (trylek) [inline src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs]: LGTM, composite images are not supposed to have "the one metadata" - they may end up having an array of those really in the [near] future but that hasn't happened yet.

**Review by trylek — APPROVED** [review]
LGTM, thank you!

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs (hunk 2 above)
Comment (fadimounir) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: Do you mean a reference to the code in `CEEInfo::canInline`? On the crossgen1 side it's doing something similar, not through a bit that we set.

Comment (trylek) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: No worries, just ignore this if it's not an obvious 1:1 correspondence, I'm just trying to preserve a bit of "history" when it comes to replacing Crossgen1 logic with Crossgen2.

File: src/coreclr/src/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs [hunk 4]
```diff
@@ -407,8 +407,6 @@ private static void FindBaseUnificationGroup(MetadataType currentType, Unificati
                 FindBaseUnificationGroup(baseType, unificationGroup);
             }
 
-            Debug.Assert(unificationGroup.IsInGroupOrIsDefiningSlot(originalDefiningMethod));
```
Comment (davidwrighton) [inline src/coreclr/src/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs]: Ugh, sadly I don't think removing this assert is correct to fix this problem. I believe the algorithm is being a little too enthusiastic about removing methods from the group. The problem I'm concerned about is if there is a 3rd type which derives. It should be able to override either the initial method, or the .override method via name/sig and redirect both methods. I've been debugging this for a while to come up with a suggestion for you here, and I'll need to continue tomorrow.

**Review by davidwrighton — CHANGES_REQUESTED** [review]
I don't believe the assertion removal is correct. It clearly stops the assert, but I believe it indicates that the algorithm is behaving incorrectly.

File: src/coreclr/src/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs (hunk 4 above)
Comment (davidwrighton) [reply src/coreclr/src/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs]: Alright, well, I think the problem is that the logic for separating from the group at line 422 shouldn't be
```
if (nameSigMatchMemberMethod != null)
```
But instead I think it should be 
```
if ((nameSigMatchMemberMethod != null) && (nameSigMatchMemberMethod != memberMethod))
```
The logic is looking for an override, not an exact match. … I think. This code is as confusing as I remember it being.

Comment (fadimounir) [reply src/coreclr/src/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs]: Thank you David. I looked at your proposed fixed, and debugged a bit, and I do agree with your analysis. I'll post an updated to the PR

**Review by davidwrighton — APPROVED** [review]
That looks good

**Pull request merged**